[WIP] DX-26191: Implement alternate Store for Nessie using MongoDB

stevelorddremio opened the pull request:
This PR currently introduces the MongoDB Maven module and creates a skeleton for implementation of the Store interface.

stevelorddremio committed:
[WIP] DX-26191: Implement alternate Store for Nessie using MongoDB

stevelorddremio committed:
[WIP] DX-26191: Implement alternate Store for Nessie using MongoDB

stevelorddremio committed:
[WIP] DX-26191: Implement alternate Store for Nessie using MongoDB

stevelorddremio committed:
[WIP] DX-26191: Implement alternate Store for Nessie using MongoDB

kylep-dremio commented on versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStoreManager.java:
```diff
@@ -0,0 +1,94 @@
+import com.google.common.collect.ListMultimap;
+
+
+public class MongoDbStoreManager implements Store {
```
Please use JavaDocs liberally.

stevelorddremio replied:
Added JavaDocs.

stevelorddremio committed:
[WIP] DX-26191: Rename MongoDbStore.java to drop Manager from name. Add JavaDocs.

keeratsingh commented on versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  /**
+   * Creates a connection to an existing database or creates the database if it does not exist.
```
I don't believe this is totally accurate, GetDatabase doesn't really create a Database, it only returns a MongoDatabase representing the specified database.
The database is lazily created when an operation like a create collection is called.
http://mongodb.github.io/mongo-java-driver/3.7/javadoc/com/mongodb/MongoClient.html#getDatabase-java.lang.String-

stevelorddremio replied:
Fair point. I will update the wording.

stevelorddremio replied:
Updated wording.

stevelorddremio committed:
[WIP] DX-26191: Add implementation for L2 puts to MongoDB

keeratsingh commented on versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,141 @@
+    if (mongoClient != null) {
+      mongoClient.close();
+      mongoClient = null;
+    }
```
MongoClient implements the Closeable and AutoCloseable interfaces, why not use those instead of calling close explicitly? Or why not use Try-with-resource, it should automatically close the MongoClient once it is out of scope.

stevelorddremio replied:
Is this not calling the close() method defined by AutoCloseable?

stevelorddremio replied:
I see what you mean. I now use try-with-resource.

kylepbit replied:
try-with-resource doesn't appear to make sense here - there is an explicit close and try-with-resource is for function or block scoped items. The reference to AutoCloseables.close() is an reference to a utility method which is not exposed here, given that I would say that this is fine.

stevelorddremio replied:
I did re-implement using try-with-resource. Given that I need the MongoClient for the methods that have not been implemented yet, I'm starting to see that try-with-resource makes it cumbersome as I keep having to instantiate the resource for each method. Sorry @keeratsingh I'll revert this change and go with @kylep-dremio suggestion.

stevelorddremio committed:
[WIP] DX-26191: Add Nessie utility class to generate test ValueType entities and Codecs for ValueType conversions

tifflhl commented on pom.xml:
```diff
@@ -461,6 +462,11 @@
       </dependency>
+      <dependency>
+        <groupId>org.mongodb</groupId>
+        <artifactId>mongodb-driver-sync</artifactId>
```
Why the synchronous driver instead of the asynchronous one?

stevelorddremio replied:
The async driver is called mongodb-driver-reactivestreams. I need to investigate the benefits of each, unless you have a particular view?

kylepbit replied:
There appears to be little difference - just as long as this is a conscious decision and there is no reason to use async.

stevelorddremio replied:
At the moment I haven't found use of Subscriber pattern in Nessie that would warrant using mongodb-driver-reactivestreams. I might have missed it. So using the sync driver would fit the bill.

tifflhl commented on versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,141 @@
+
+  @Override
+  public <V> boolean putIfAbsent(ValueType type, V value) {
+    return false;
```
Are these stubbed implementations or unsupported operations? Better to throw an error if they are unsupported.

csolonynka replied:
If these are stubbed, then put in a comment to say that is the case.

stevelorddremio replied:
I have added throwing UnsupportedOperationException() exception to all unimplemented methods.

stevelorddremio committed:
[WIP] DX-26191: Updated Mongo implementation from PR comment.

stevelorddremio committed:
[WIP] Updated Mongo implementation to serialize L2 to BSON

stevelorddremio committed:
[WIP] DX-26191: Updated Mongo implementation from PR comment.

kylep-dremio committed:
Add codecs for L1, L3, Fragment.
Clean up code and eliminate build errors.

kylep-dremio committed:
Add samples for L1, L2, L3 objects.
Clean up additional portions of code.
Simplify Codecs.

kylep-dremio committed:
Hook up the MongoDB configuration into the Quarkus server.
Change MongoDbStore to take a connection string to allow configuration.
Fix some warnings.

kylep-dremio committed:
Correct a few code style issues.

kylep-dremio committed:
Add deserialization of BSON to Entity.
Add unit tests for put of L1 and L3.
Correct some issues with encode/decode of NUMBER, STRING_SET.

stevelorddremio committed:
Add MongoDB loadSingle() implementation

stevelorddremio committed:
MongoDB implementation of Store.loadSingle() corrected to find correct key type.

kylep-dremio committed:
Add the remaining sample entities.
Add missing codecs for the entities.
Move CodecProvider to the entity package to
allow it to have access to the different entity types
to allow the codecs to be properly detected.

kylep-dremio committed:
Add flapdoodle to unit test with mongo.

kylep-dremio committed:
Correct checkstyle issues.

kylep-dremio committed:
Correct loadSingle behaviour.
Add copied implementation of putIfAbsent.
Add a getRefs implementation.

kylep-dremio committed:
Fixed checkstyle warnings.
Same minor refactoring to reduce duplication.

kylep-dremio committed:
Refactoring to simplify test code and increase coverage.

kylep-dremio committed:
Fix checkstyle issue.

kylep-dremio committed:
Add additional tests around encode/decode.

stevelorddremio committed:
Add scripts to configure test environment for sharding in MongoDB and add scripts to set up sharding.

kylep-dremio committed:
Merge commit '9ce6d6e610bff29a2cfb27d577063c9ca9e84a7a' into DX-26191

kylep-dremio committed:
Add an implementation of save() along with test.

kylep-dremio committed:
Fix checkstyle issues.

stevelorddremio committed:
Updated Mongo collection names to be more meaningful.

kylep-dremio committed:
Correct save with transaction.

kylep-dremio committed:
Remove configuration of MongoDB from Quarkus.

To be re-added once MongoDB implementation complete.

stevelorddremio committed:
Update scripts to shard MongoDB instances running on localhost

kylep-dremio committed:
Respond to some of the code review comments.

kylep-dremio committed:
Change to use the reactive streams mongo client.
Move SampleEntities to a common package (not to a common project yet).
Note that L1/L3 constructor visibility was slightly relaxed to allow construction.
Simplify common TestStore code.
Add proper MongoDB putIfAbsent and put behaviour.

kylep-dremio committed:
Randomize the strings used in the sample Entities.

kylep-dremio committed:
Add better test coverage.

kylep-dremio committed:
Merge branch 'main' into DX-26191

kylep-dremio committed:
Merge branch 'main' into DX-26191

kylep-dremio committed:
Remove CodecProvider dependency on Internal classes.
Move AbstractTestStore to better package, to be relocated to
versioned/tests once that won't introduce a circular dependency.
Some cleanup of the code.

kylep-dremio committed:
Move MongoStoreConfig to the mongo package,
where it should have been.

kylep-dremio committed:
Reduce visibility of many methods & classes.

stevelorddremio committed:
Add sharding to flapdoodle for MongoDB backing store

kylep-dremio committed:
Remove lingering pom reference to Mongo in Quarkus.

To be re-added once Mongo is enabled.

stevelorddremio committed:
Refactor of test MongoDB sharding implementation.

stevelorddremio committed:
MongoDB replace static methods with singleton classes. Optimize collection sharding.

kylep-dremio committed:
Remove old mongo pom artifacts, use correct version property.

stevelorddremio committed:
MongoDB removed programmatic sharding of flapdoodle for testing until fixed.

stevelorddremio committed:
MongoDB comment update

kylep-dremio committed:
Addressing code review comments.

kylep-dremio committed:
Fix a randomization issue which caused all entities to use the same values.

kylep-dremio committed:
Small addition of clarifying comment.

kylep-dremio committed:
Code review comments.
